primitives/utils/src/metrics.rs

```diff
@@ -0,0 +1,147 @@
+
+use lazy_static::lazy_static;
+use prometheus::{
+    Registry, Error as PrometheusError, Opts,
+    core::{
+        GenericGaugeVec, AtomicU64,
+        GenericGauge,
+	}
+};
+use std::time::Instant;
+use std::convert::TryFrom;
+use parking_lot::Mutex;
+
+use crate::channels::{ Sender, Receiver, OnFullStrategy, channel_with_strategy };
+
+lazy_static! {
+
+    pub static ref TOKIO_THREADS_TOTAL: GenericGauge<AtomicU64> = GenericGauge::new(
+        "tokio_threads_total", "Total number of threads created"
+    ).expect("Creating of statics doesn't fail. qed");
+
+    pub static ref TOKIO_THREADS_ALIVE: GenericGauge<AtomicU64> = GenericGauge::new(
+        "tokio_threads_alive", "Number of threads alive right now"
+    ).expect("Creating of statics doesn't fail. qed");
+
+	pub static ref UNBOUNDED_CHANNELS_COUNTER : GenericGaugeVec<AtomicU64> = GenericGaugeVec::new(
+        Opts::new("internals_unbounded_channels", "items in each mpsc::unbounded instance"),
+        &["entity"]
+    ).expect("Creating of statics doesn't fail. qed");
+
+    pub static ref BLOCK_IMPORT: TimeSeriesCollector = TimeSeriesCollector::new();
+}
+
+lazy_static! {
+    // Non public ones
+
+	static ref BLOCK_IMPORT_INNER : GenericGaugeVec<AtomicU64> = GenericGaugeVec::new(
+        Opts::new("block_import", "Block Import"),
+        &["subtype"]
+    ).expect("Creating of statics doesn't fail. qed");
+
+}
+
+struct TimeSeriesInfo {
+	count: u64,
+	lower_median: u64,
+	median: u64,
+	higher_median: u64,
+	average: u64
+}
+
+
+impl From<Vec<u64>> for TimeSeriesInfo {
+	fn from(mut input: Vec<u64>) -> Self {
+		let count = input.len();
+		if let Some(only_value) = match count {
+			0 => Some(0),
+			1 => Some(input[0]),
+			_ => None
+		} {
+			return TimeSeriesInfo {
+				count: u64::try_from(count).expect("Usize always fits into u64. qed"),
+				lower_median: only_value,
+				median: only_value,
+				higher_median: only_value,
+				average: only_value,
+			}
+		}
+
+		input.sort();
+		let median_pos = count.div_euclid(2);
+		let median_dif = median_pos.div_euclid(2);
+		let count = u64::try_from(count).expect("Usize always fits into u64. qed");
+		let average = input.iter().fold(0u64, |acc, val| acc + val).div_euclid(count);
+
+		TimeSeriesInfo {
+			count,
+			lower_median: input[median_pos - median_dif],
+			median: input[median_pos],
```

Comment:
Say we would later on want to accumulate this. Correct me if I am wrong, but as far as I know medians are not associative. Thus accumulation would be impossible.

E.g. let's say we have the events `[1, 2, 3, 4]`. If they arrive all at once we get an average of `2.5`. If first 3 arrive and then one we get an average of `((1+2+3)÷3 + 4÷1)÷2 = 3`.

Reply:
average != median. I am recording median, lower and upper median together with average and total for a reason – they only make sense in context. Just one block (say an upgrade or migration) would easily put a huge spike on all of these. With the context of count and alike you have some more informed over-time view on this.

While I think you technically can aggregate over medians (within some limits), I am not sure whether accumulation makes sense here at all as this is highly hardware, load and environmentally dependent. This is used primarily for seeing an overall trend (like it growing over time or being largely different after restarts) or notice difference between implementations (it was a magnitude different between two versions), but is less informative once you start to aggregate.